apex/amp/utils.py

```diff
@@ -86,13 +86,20 @@ def cached_cast(cast_fn, x, cache):
         return type(x)([cached_cast(y) for y in x])
     if x in cache:
         cached_x = cache[x]
-        # During eval, it's possible to end up caching casted weights
-        # with requires_grad == False. This is then a problem when they
-        # get reused on the next train iter. So we ensure that cached
-        # weights have same requires_grad flag of most recent request.
-        if x.requires_grad != cached_x.requires_grad:
-            cached_x.requires_grad_(x.requires_grad)
-        return cache[x]
+        if x.requires_grad and cached_x.requires_grad:
+            # Make sure x is actually cached_x's autograd parent.
+            if cached_x.grad_fn.next_functions[1][0].variable is not x:
+                raise RuntimeError("x and cache[x] both require grad, but x is not "
+                                   "cache[x]'s parent.  This is likely an error.")
+        # During eval, it's possible to end up caching casted weights with
+        # requires_grad=False.  On the next training iter, if cached_x is found
+        # and reused from the cache, it will not actually have x as its parent.
+        # Therefore, we choose to invalidate the cache (and force refreshing the cast)
+        # if x.requires_grad and cached_x.requires_grad do not match.
+        if torch.is_grad_enabled() and x.requires_grad != cached_x.requires_grad:
```

Comment:
Re: our slack convo, could you add one sentence explaining why the `torch.is_grad_enabled()` check matters?